### PR TITLE
Feature/Custom webpack extract-i18n Builder Support

### PR DESCRIPTION
**Pull request by patnolanireland**
## Overview

The [@angular-devkit/build-angular](https://github.com/angular/angular-cli/tree/master/packages/angular_devkit/build_angular) has a build target for `extract-i18n`.  This builder is currently missing from this repo.

This PR simply adds the missing builder from https://github.com/angular/angular-cli/tree/master/packages/angular_devkit/build_angular/src/extract-i18n in the same fashion as the existing supported builders.

The code has been branched from the `7.x.x` branch and I've not looked at merging it into master yet.  Happy to do so if need be, just needed support for Angular 7 as a matter of urgency.  I've bumped the `package.json` to `7.4.4` so that a new version can be deployed by building, running the `ci` script to package and then publish to npm.

### CODE REVIEW

File: packages/custom-webpack/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@angular-builders/custom-webpack",
-  "version": "7.4.2",
+  "version": "7.4.4",
```
Comment (meltedspark) [inline packages/custom-webpack/package.json]: This is a patch bump, you need to bump a minor version (new feature).

File: packages/custom-webpack/src/extract-i18n/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/**
```
Comment (meltedspark) [inline packages/custom-webpack/src/extract-i18n/index.d.ts]: No need in this file (unless you tell me there is an explicit need)

**Comment by meltedspark** [comment]
As for master, I'm afraid you won't be able to merge this one directly to master - too many differences between 7 and 8. But you can create a separate PR for version 8. There are plenty of examples of how things work now in master so it should be enough for you I believe. If you have any questions - don't hesitate to ask.

File: packages/custom-webpack/package.json (hunk 1 above)
Comment (patnolanireland) [reply packages/custom-webpack/package.json]: No worries, updated to `7.5.0`

File: packages/custom-webpack/src/extract-i18n/index.d.ts (hunk 2 above)
Comment (patnolanireland) [reply packages/custom-webpack/src/extract-i18n/index.d.ts]: I doubled checked an you're correct, file removed.

**Comment by patnolanireland** [comment]
> Thanks for your contribution Pat!

No worries @meltedspark you're welcome.  

> I left a few comments, please address them.

I've addressed the comments and pushed the changes.  I left a message on the Slack channel just after creating the PR but it doesn't look like it's very active.  

> * Could you please elaborate on your use case? What exactly are you changing in the build during the extraction? It could be also a great example in the README instead of the standard one.

Essentially the Use Case for this functionality came about as a result of integrating in two libraries into a project I'm currently working on, namely

1. js-to-styles-var-loader
2. ngx-i18nsupport

The first library requires requires the following extra-webpack.config.js

```
const path = require('path');

module.exports = {
  module: {
    rules: [
      {
        test: /\.less$/,
        use: [
          {
            loader: 'js-to-styles-var-loader',
          },
        ],
      },
      {
        test: /\.scss$/,
        use: [
          {
            loader: 'js-to-styles-var-loader',
          },
        ],
      },
    ],
  },
};
```

Adding the ngx-i18nsupport library adds an entry to the `scripts` object in the package.json e.g.

```
"scripts": {
    "extract-i18n": "ng xi18n my-cool-app --i18n-format xlf --output-path i18n --i18n-locale en && ng run my-cool-app:xliffmerge",

}
```

Out of the box this leverages the `extract-i18n` builder from the Angular Dev Kit `@angular-devkit/build-angular` which doesn't know about the custom webpack config and the build breaks as the standard build doesn't know about the `js-to-styles-var-loader`.

Happy to put a blurb into the `README.md` to convey the above Use Case.  Let me know if you're good with this.

> * Have you checked that it works? If no, please do. If yes, please add an extraction run to the ci.sh script to make sure it works in every build.

Yes.  I've tested locally by `yarn link` and also as a sanity test I've built and deployed to a separate Org Scoped Package entitled @delta-ready-tech-angular-builders/custom-webpack and have npm installed that into the repo I'm working on.  Feel free to take it for a spin if you wish.  Also happy to take down that published package if this PR gets merged.

**Comment by patnolanireland** [comment]
> As for master, I'm afraid you won't be able to merge this one directly to master - too many differences between 7 and 8. But you can create a separate PR for version 8. There are plenty of examples of how things work now in master so it should be enough for you I believe. If you have any questions - don't hesitate to ask.

Yeah I noticed that `master` is using the RC of version 8.  Happy to create a separate PR for version 8, might be a few days before I get a chance to tend to it though.

**Comment by meltedspark** [comment]
@patnolanireland Hey how it's going? Do you plan to complete it or you'd like me to take it from here?

**Comment by patnolanireland** [comment]
> @patnolanireland Hey how it's going? Do you plan to complete it or you'd like me to take it from here?

Hey @meltedspark, apologies for the delay in getting back to you I've got swamped with work and personal life and as such haven't had a chance to spend the time getting the ci test in that you've requested.  Hopefully I'll get some time to finish off the 7.x.x branch this weekend.  If I don't manage to do so happy for you to take it from here.

**Comment by meltedspark** [comment]
No worries take your time.

File: packages/custom-webpack/README.md [hunk 3]
```diff
@@ -195,33 +195,20 @@ Builder options:
...
-    ],
-  },
+  plugins: [
+    new I18nXlfAnnotateAppVersionPlugin()
+  ]
 };
 ```
 
-In this example `module.rules` entry from `extra-webpack.config.js` will be appended to `module.rules` entry from Angular CLI underlying webpack config forcing these loaders to run first, before any of the original loaders defined by the Angular CLI.
+In this example `module.rules` entry from `extra-webpack.config.js` will be appended to `module.rules` entry from Angular CLI's underlying webpack config.
```
Comment (meltedspark) [inline packages/custom-webpack/README.md]: Please update it with description that is relevant to the example

File: packages/custom-webpack/examples/append-webpack-plugins/angular.json [hunk 4]
```diff
@@ -98,6 +124,21 @@
               "**/node_modules/**"
             ]
           }
+        },
+        "xliffmerge": {
```
Comment (meltedspark) [inline packages/custom-webpack/examples/append-webpack-plugins/angular.json]: I'd really prefer to keep this example as simple as possible. As far as I understand `xliffmerge` is not necessary (as it runs after the webpack plugin was run) an it's enough to verify the file exists after simple `xi18n` run. Am I missing something? If not please remove it as it complicates the example.

Comment (patnolanireland) [reply packages/custom-webpack/examples/append-webpack-plugins/angular.json]: >As far as I understand xliffmerge is not necessary (as it runs after the webpack plugin was run) an it's enough to verify the file exists after simple xi18n run.

100% correct, removed as requested.

File: packages/custom-webpack/README.md (hunk 3 above)
Comment (patnolanireland) [reply packages/custom-webpack/README.md]: Updated to reflect the custom plugin in the webpack config above

**Comment by patnolanireland** [comment]
> Hey @patnolanireland, really great job, thank you so much!
> I left a few comments, please address them.
> Also, (just verifying) the `xi18n` run fails if you run it with the default Angular builder, right? Because this is what will actually indicate problems.

🆒 no worries 🥂 The default builder won't always fail unless the custom webpack configuration used by that builder causes it to e.g. if you have a loader that is required in the custom config to yield a successful build. 

So basically I could have got around my issue by creating a custom browser build target in my local project for i18n that didn't use the custom webpack config.  The sample plugin illustrates by way of an albeit contrived example a case where the `extract-i18n` custom builder can be used to modify the extraction process which wouldn't be possible without this PR.

**Comment by meltedspark** [comment]
@patnolanireland Do you plan to add this feature to version 8?

**Comment by patnolanireland** [comment]
> @patnolanireland Do you plan to add this feature to version 8?

Yep sure do @meltedspark.  I've not looked into it yet though.  Is there anything specific that has changed from an implementation perspective or is it just a matter of extracting the code from the 7.x.x branch into a branch from master and raising a PR?  If I remember correctly there has been a significant amount of restructure in the repo that would make merging 7.x.x into master more cumbersome than just branching off master and porting the changes over.  Happy to be directed towards any docs or resort to reading the Angular CLI source if necessary.

**Comment by meltedspark** [comment]
Unfortunately you won't be able to just cherry-pick the code, because the internal API has changed. I also changed the CI setup a bit, incorporated lerna etc.  
You can start by reading the [updated article](https://blog.angularindepth.com/angular-cli-under-the-hood-builders-demystified-v2-e73ee0f2d811) about builders and then port your changes to the new version.

**Comment by meltedspark** [comment]
If you have any further questions find me in Slack.

**Pull request merged**